File: 6. PCA Image Compression.py
# Topic: Data compression via PCA — compress and reconstruct small image using numpy SVD

import numpy as np

def compress_image_pca(img, k):
    # img: H x W grayscale
    U, S, Vt = np.linalg.svd(img, full_matrices=False)
    S_k = np.diag(S[:k])
    rec = U[:,:k] @ S_k @ Vt[:k,:]
    err = np.linalg.norm(img - rec)
    return rec, err

if __name__ == '__main__':
    img = np.array([[1,2,3],[4,5,6],[7,8,9]], dtype=float)
    rec, err = compress_image_pca(img, k=1)
    print('Reconstruction error (k=1):', err)

# Time Complexity: O(min(H,W)^2 * max(H,W))
# Space Complexity: O(H*W)

File: 6. PCA Image Compression.cpp
// Topic: Data compression via PCA (image as matrix) — demo using SVD via simple numeric method
#include <bits/stdc++.h>
using namespace std;

// For simplicity, we'll perform SVD via power method on covariance — small educational demo
int main()
{
    vector<vector<double>> img = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};  // small 3x3 "image"
    // In practice use SVD (Eigen or LAPACK). Here we note process and skip heavy implementation.
    cout << "This demo is a placeholder: use SVD (Eigen or LAPACK) to compress/reconstruct images."
         << endl;
    return 0;
}

// Time Complexity: dependent on SVD
// Space Complexity: dependent on SVD
